src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -103,9 +127,47 @@ public void launch() {
             kernel.getContext().get(c).startup();
         }
 
+        final Queue<DeviceIdentityInterface> provisioningPlugins = findProvisioningPlugins();
         // Must be called before everything else so that these are available to be
         // referenced by main/dependencies of main
         final Queue<String> autostart = findBuiltInServicesAndPlugins(); //NOPMD
+        // run the provisioning if device is not provisioned
+        if (!kernel.getContext().get(DeviceConfiguration.class).isDeviceConfiguredToTalkToCloud()
+                && !provisioningPlugins.isEmpty()) {
+            logger.atDebug().log("Found provisioning plugins to run");
+            RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                    .maxAttempt(MAX_PROVISIONING_PLUGIN_RETRY_ATTEMPTS)
+                    .retryableExceptions(Collections.singletonList(RetryableProvisioningException.class))
+                    .build();
+            ExecutorService executorService = kernel.getContext().get(ExecutorService.class);
+            executorService.execute(() -> {
+                provisioningPlugins.stream().forEach((plugin) -> {
```

Comment:
Also, is there a test for this?

Reply:
There are no chains but if multiple plugins want to update different parts of config then they can. Ordering of plugins is out of scope of this work.

Reply:
Makes sense that it's out of scope but I'm a bit concerned with the fact that having multiple plugins introduces the possibility of a plugin executing after some other plugin inadvertently overwriting config set by the former plugin. I imagine it will also be difficult for customers to debug

Reply:
As per discussed design, we are going to allow customers to specify a trusted plugins jar using easy setup. Easy setup will then copy that jar to trusted plugins directory. If we were to restrict it to one plugin, it will be difficult to restrict customers to put only one provision plugin jar in the trusted plugins directory OR to put only one DeviceIdentityInterface implementing class in a single jar. If such a scenario happens then nucleus will face a problem of which plugin to pick. 
In the current impl, nucleus does not have to choose. We still depend on customers to configure the plugins correctly. For debugging, we have logs at places to show which plugin is executing.

Reply:
As discussed, we will error out if multiple provisioning plugins are found. There is no compelling use case for multiple provisioning plugins right now.